# Long_read_genome_assembly
This is a compilation of the scripts I used for genome assembly using long read sequencing data (ONT)
Workflow
1. Basecalling (Guppy)
2. Checking the quality of our data (pycoQC)
3. Filtering reads (Nanolyse)
4. Removing adapters (Porechop)
5. Quality check (nanoQC)
6. Genome assembler

    6.1 Flye
    
    6.2 Canu
    
    6.3 Redbean
7. Assembly polishing with Racon




## 1. Basecaling (Guppy)
We sequenced the genome using minion without live basecalling. So we are going to use guppy (v5.0.7) on GPU (In Nesi) for basecalling, we will use disable_qscore_filtering becasue we don't want to separate reads based on their quality scores in to pass and fail folders instead get all the fastq files in one folder and then decide on quality filtering.

Here is the script to [run](guppy.sl)

```
sbatch guppy.sl
```
## 2. Checking the quality of our data (pycoQC)
We can check the quality of our data using pycoQC, we can install it using miniconda, lets activate our miniconda using `conda activate` from `miniconda2/bin` directory

```
conda create -n pycoQC python=3.6
# or just
conda install -c bioconda pycoqc # pycoQC was built in python3 so need python3 environment.
```
pycoQC uses `sequencing_summary.txt` generated by guppy or other basecaller as an input. We can use script below to run pycoQC

```
#!/bin/bash -e

#SBATCH --nodes 1
#SBATCH --cpus-per-task 1
#SBATCH --ntasks 10
#SBATCH --partition=large
#SBATCH --job-name pycoqc
#SBATCH --mem=50G
#SBATCH --time=00:20:00
#SBATCH --account=uoo02752
#SBATCH --output=%x.%j.out
#SBATCH --error=%x.%j.err
#SBATCH --mail-type=ALL
#SBATCH --mail-user=bhaup057@student.otago.ac.nz
#SBATCH --hint=nomultithread

export PATH="/nesi/nobackup/uoo02752/nematode/bin/miniconda3/bin:$PATH"

pycoQC -f ../../sequencing_summary.txt -o pycoQC_output.html
```
PycoQC produces an interactive html report, below is a screenshot for our data:
![PycoQC Report]
(<img width="866" alt="Screen Shot 2021-06-11 at 2 42 05 PM" src="https://user-images.githubusercontent.com/49458562/121623142-4df0eb00-cac3-11eb-91e0-5466bccb7482.png">
)

## 3. Filtering reads (Nanolyse)
We have used `DNACS` during nanopore sequencing library preparation so we will use `Nanolyse` to remove lamda DNA from our fastq files
We can install Nanolyse using miniconda environment, see [Nanolyse page on how](https://github.com/wdecoster/NanoLyse)
Let's run below script to process our data with nanolyse.

```
#!/bin/bash -e

#SBATCH --nodes 1
#SBATCH --cpus-per-task 1
#SBATCH --ntasks 10
#SBATCH --partition=large
#SBATCH --job-name nanolyse.job
#SBATCH --mem=50G
#SBATCH --time=08:00:00
#SBATCH --account=uoo02752
#SBATCH --output=%x.%j.out
#SBATCH --error=%x.%j.err
#SBATCH --mail-type=ALL
#SBATCH --mail-user=bhaup057@student.otago.ac.nz
#SBATCH --hint=nomultithread

export PATH="/nesi/nobackup/uoo02752/nematode/bin/miniconda3/bin:$PATH"

cat Nem.ont.merged.fastq | NanoLyse | gzip > m.neg_nanopore_filtered.fastq.gz
```
This will remove all the lamda DNA in our data
I have had problem running Nanolyse: that it couldn't find the lamdaDNA file `dna_cs.fasta` to filter out, in that case make a text file `dna_cs.fasta` in the current working directory with lamda dna sequence (below) and slightly change above script as:

```
cat Nem.ont.merged.fastq | NanoLyse --reference ./dna_cs.fasta | gzip > m.neg_nanopore_filtered.fastq.gz
```
dna_cs.fasta

```
>DNA_CS
GCCATCAGATTGTGTTTGTTAGTCGCTTTTTTTTTTTGGAATTTTTTTTTTGGAATTTTTTTTTTGCGCTAACAACCTCCTGCCGTTTTGCCCGTGCATATCGGTCACGAACAAATCTGATTACTAAACACAGTAGCCTGGATTTGTTCTATCAGTAATCGACCTTATTCCTAATTAAATAGAGCAAATCCCCTTATTGGGGGTAAGACATGAAGATGCCAGAAAAACATGACCTGTTGGCCGCCATTCTCGCGGCAAAGGAACAAGGCATCGGGGCAATCCTTGCGTTTGCAATGGCGTACCTTCGCGGCAGATATAATGGCGGTGCGTTTACAAAAACAGTAATCGACGCAACGATGTGCGCCATTATCGCCTAGTTCATTCGTGACCTTCTCGACTTCGCCGGACTAAGTAGCAATCTCGCTTATATAACGAGCGTGTTTATCGGCTACATCGGTACTGACTCGATTGGTTCGCTTATCAAACGCTTCGCTGCTAAAAAAGCCGGAGTAGAAGATGGTAGAAATCAATAATCAACGTAAGGCGTTCCTCGATATGCTGGCGTGGTCGGAGGGAACTGATAACGGACGTCAGAAAACCAGAAATCATGGTTATGACGTCATTGTAGGCGGAGAGCTATTTACTGATTACTCCGATCACCCTCGCAAACTTGTCACGCTAAACCCAAAACTCAAATCAACAGGCGCCGGACGCTACCAGCTTCTTTCCCGTTGGTGGGATGCCTACCGCAAGCAGCTTGGCCTGAAAGACTTCTCTCCGAAAAGTCAGGACGCTGTGGCATTGCAGCAGATTAAGGAGCGTGGCGCTTTACCTATGATTGATCGTGGTGATATCCGTCAGGCAATCGACCGTTGCAGCAATATCTGGGCTTCACTGCCGGGCGCTGGTTATGGTCAGTTCGAGCATAAGGCTGACAGCCTGATTGCAAAATTCAAAGAAGCGGGCGGAACGGTCAGAGAGATTGATGTATGAGCAGAGTCACCGCGATTATCTCCGCTCTGGTTATCTGCATCATCGTCTGCCTGTCATGGGCTGTTAATCATTACCGTGATAACGCCATTACCTACAAAGCCCAGCGCGACAAAAATGCCAGAGAACTGAAGCTGGCGAACGCGGCAATTACTGACATGCAGATGCGTCAGCGTGATGTTGCTGCGCTCGATGCAAAATACACGAAGGAGTTAGCTGATGCTAAAGCTGAAAATGATGCTCTGCGTGATGATGTTGCCGCTGGTCGTCGTCGGTTGCACATCAAAGCAGTCTGTCAGTCAGTGCGTGAAGCCACCACCGCCTCCGGCGTGGATAATGCAGCCTCCCCCCGACTGGCAGACACCGCTGAACGGGATTATTTCACCCTCAGAGAGAGGCTGATCACTATGCAAAAACAACTGGAAGGAACCCAGAAGTATATTAATGAGCAGTGCAGATAGAGTTGCCCATATCGATGGGCAACTCATGCAATTATTGTGAGCAATACACACGCGCTTCCAGCGGAGTATAAATGCCTAAAGTAATAAAACCGAGCAATCCATTTACGAATGTTTGCTGGGTTTCTGTTTTAACAACATTTTCTGCGCCGCCACAAATTTTGGCTGCATCGACAGTTTTCTTCTGCCCAATTCCAGAAACGAAGAAATGATGGGTGATGGTTTCCTTTGGTGCTACTGCTGCCGGTTTGTTTTGAACAGTAAACGTCTGTTGAGCACATCCTGTAATAAGCAGGGCCAGCGCAGTAGCGAGTAGCATTTTTTTCATGGTGTTATTCCCGATGCTTTTTGAAGTTCGCAGAATCGTATGTGTAGAAAATTAAACAAACCCTAAACAATGAGTTGAAATTTCATATTGTTAATATTTATTAATGTATGTCAGGTGCGATGAATCGTCATTGTATTCCCGGATTAACTATGTCCACAGCCCTGACGGGGAACTTCTCTGCGGGAGTGTCCGGGAATAATTAAAACGATGCACACAGGGTTTAGCGCGTACACGTATTGCATTATGCCAACGCCCCGGTGCTGACACGGAAGAAACCGGACGTTATGATTTAGCGTGGAAAGATTTGTGTAGTGTTCTGAATGCTCTCAGTAAATAGTAATGAATTATCAAAGGTATAGTAATATCTTTTATGTTCATGGATATTTGTAACCCATCGGAAAACTCCTGCTTTAGCAAGATTTTCCCTGTATTGCTGAAATGTGATTTCTCTTGATTTCAACCTATCATAGGACGTTTCTATAAGATGCGTGTTTCTTGAGAATTTAACATTTACAACCTTTTTAAGTCCTTTTATTAACACGGTGTTATCGTTTTCTAACACGATGTGAATATTATCTGTGGCTAGATAGTAAATATAATGTGAGACGTTGTGACGTTTTAGTTCAGAATAAAACAATTCACAGTCTAAATCTTTTCGCACTTGATCGAATATTTCTTTAAAAATGGCAACCTGAGCCATTGGTAAAACCTTCCATGTGATACGAGGGCGCGTAGTTTGCATTATCGTTTTTATCGTTTCAATCTGGTCTGACCTCCTTGTGTTTTGTTGATGATTTATGTCAAATATTAGGAATGTTTTCACTTAATAGTATTGGTTGCGTAACAAAGTGCGGTCCTGCTGGCATTCTGGAGGGAAATACAACCGACAGATGTATGTAAGGCCAACGTGCTCAAATCTTCATACAGAAAGATTTGAAGTAATATTTTAACCGCTAGATGAAGAGCAAGCGCATGGAGCGACAAAATGAATAAAGAACAATCTGCTGATGATCCCTCCGTGGATCTGATTCGTGTAAAAAATATGCTTAATAGCACCATTTCTATGAGTTACCCTGATGTTGTAATTGCATGTATAGAACATAAGGTGTCTCTGGAAGCATTCAGAGCAATTGAGGCAGCGTTGGTGAAGCACGATAATAATATGAAGGATTATTCCCTGGTGGTTGACTGATCACCATAACTGCTAATCATTCAAACTATTTAGTCTGTGACAGAGCCAACACGCAGTCTGTCACTGTCAGGAAAGTGGTAAAACTGCAACTCAATTACTGCAATGCCCTCGTAATTAAGTGAATTTACAATATCGTCCTGTTCGGAGGGAAGAACGCGGGATGTTCATTCTTCATCACTTTTAATTGATGTATATGCTCTCTTTTCTGACGTTAGTCTCCGACGGCAGGCTTCAATGACCCAGGCTGAGAAATTCCCGGACCCTTTTTGCTCAAGAGCGATGTTAATTTGTTCAATCATTTGGTTAGGAAAGCGGATGTTGCGGGTTGTTGTTCTGCGGGTTCTGTTCTTCGTTGACATGAGGTTGCCCCGTATTCAGTGTCGCTGATTTGTATTGTCTGAAGTTGTTTTTACGTTAAGTTGATGCAGATCAATTAATACGATACCTGCGTCATAATTGATTATTTGACGTGGTTTGATGGCCTCCACGCACGTTGTGATATGTAGATGATAATCATTATCACTTTACGGGTCCTTTCCGGTGAAAAAAAAGGTACCAAAAAAAACATCGTCGTGAGTAGTGAACCGTAAGC
```
## 4. Removing adapters (Porechop)
We can use Porechop to remove adapters. Adapters on the ends are trimmed off and if on middle, they are treated as chimeric and chopped into seperate reads [Porechop](https://github.com/rrwick/Porechop#quick-usage-examples)
We used a flag `--trim_barcodes` with `Guppy` during base calling so the adapters on ends of the reads are already removed, however porechop helped remove adapters from middle of the reads.
Porechop is available as a module in Nesi so no need to install it we can use it as below:

```
#!/bin/bash -e

#SBATCH --nodes 1
#SBATCH --cpus-per-task 1
#SBATCH --ntasks 10
#SBATCH --partition=large
#SBATCH --job-name porechop
#SBATCH --mem=50G
#SBATCH --time=04:00:00
#SBATCH --account=uoo02752
#SBATCH --output=%x.%j.out
#SBATCH --error=%x.%j.err
#SBATCH --mail-type=ALL
#SBATCH --mail-user=bhaup057@student.otago.ac.nz
#SBATCH --hint=nomultithread

module load Porechop/0.2.4-gimkl-2020a-Python-3.8.2

porechop -i ../m.neg_nanopore_filtered.fastq.gz -o m.neg_ont_nanolyse_porechop.fastq.gz --threads 10
```

## 5. Quality check (nanoQC)
We can install nanoQC using conda [see nanoQC](https://github.com/wdecoster/nanoQC)
Lets check the quality of our processed data

```
#!/bin/bash -e

#SBATCH --nodes 1
#SBATCH --cpus-per-task 1
#SBATCH --ntasks 10
#SBATCH --partition=large
#SBATCH --job-name nanoqc
#SBATCH --mem=50G
#SBATCH --time=02:00:00
#SBATCH --account=uoo02752
#SBATCH --output=%x.%j.out
#SBATCH --error=%x.%j.err
#SBATCH --mail-type=ALL
#SBATCH --mail-user=bhaup057@student.otago.ac.nz
#SBATCH --hint=nomultithread

export PATH="/nesi/nobackup/uoo02752/nematode/bin/miniconda3/bin:$PATH"

nanoQC ../m.neg_ont_nanolyse_porechop.fastq.gz -o ./
```

## 6. Genome assembler
We will use flye assembler for our data. I have tried several assemblers on this data with trimmed (based on quality scores) and untrimmed input reads. Flye assembly had the highest busco scores, however it produces bigger assembly (length and number of contigs). Below are the scripts to run each of assembler.

### 6.1 Flye
Flye is available as module in `Nesi` so no need to install it. We can run it with the script below, it will also run 3 iteration of genome polishing `-i 3`

```
#!/bin/bash -e

#SBATCH --nodes 1
#SBATCH --cpus-per-task 1
#SBATCH --ntasks 10
#SBATCH --partition=large
#SBATCH --job-name flye.nematode
#SBATCH --mem=50G
#SBATCH --time=72:00:00
#SBATCH --account=uoo02752
#SBATCH --output=%x_%j.out
#SBATCH --error=%x_%j.err
#SBATCH --mail-type=ALL
#SBATCH --mail-user=bhaup057@student.otago.ac.nz
#SBATCH --hint=nomultithread

module load Flye/2.7.1-gimkl-2020a-Python-2.7.18

flye --nano-raw m.neg_ont_nanolyse_porechop.fastq.gz \
--genome-size 337m -o ./M.neg_flye -t 10 -i 3
```

### 6.2 Canu
Canu is also available in Nesi as module, so no need to install it. It performs reads trimming, correction and assembly. We can either run these three steps separately or run them all together. check out [canu](https://canu.readthedocs.io/en/latest/quick-start.html) for more details.
With the script below we can run all three steps of canu together and get an assembly for our input data. Options below like `minReadLength=500 minOverlapLength=300 correctedErrorRate=0.146` are tuned for our low coverage ONT data.

```
#!/bin/bash -e

#SBATCH --nodes 1
#SBATCH --cpus-per-task 1
#SBATCH --ntasks 10
#SBATCH --partition=bigmem
#SBATCH --job-name canu
#SBATCH --mem=100G
#SBATCH --time=72:00:00
#SBATCH --account=uoo02752
#SBATCH --output=%x_%j.out
#SBATCH --error=%x_%j.err
#SBATCH --mail-type=ALL
#SBATCH --mail-user=bhaup057@student.otago.ac.nz
#SBATCH --hint=nomultithread

#module load Canu/2.0-GCC-9.2.0
module load Canu/2.1.1-GCC-9.2.0

canu \
-p m.neg -d canu.nem.337 genomeSize=337m minReadLength=500 minOverlapLength=300 correctedErrorRate=0.146 \
useGrid=true gridOptions="--time=72:00:00 --partition=bigmem --account=uoo02752 --hint=nomultithread" \
-nanopore-raw m.neg_ont_nanolyse_porechop.fastq.gz
```

### 6.3 Redbean
Redbean previously known as wtdbg2 is also available as module in `Nesi`. We will run this assembler two times with two different (kmer) settings and merge those two assemblies to get the best out of it.

Script for running first Redbean

```
#!/bin/bash -e

#SBATCH --nodes 1
#SBATCH --cpus-per-task 1
#SBATCH --ntasks 10
#SBATCH --partition=large
#SBATCH --job-name wtdbg2_2.1
#SBATCH --mem=50G
#SBATCH --time=48:00:00
#SBATCH --account=uoo02752
#SBATCH --output=%x_%j.out
#SBATCH --error=%x_%j.err
#SBATCH --mail-type=ALL
#SBATCH --mail-user=bhaup057@student.otago.ac.nz
#SBATCH --hint=nomultithread

module load wtdbg/2.5-GCC-9.2.0

wtdbg2 -x ont -g 67m -t 10 -i m.neg_ont_nanolyse_porechop.fastq.gz \
-fo m.neg_wtdbgp0k17 -p 0 -k 17 -S 2 -L 5000 --edge-min 2 —rescue-low-cov-edges -A --aln-dovetail -1     # for denovo assembly

wtpoa-cns -t 10 -i m.neg_wtdbgp0k17.ctg.lay.gz -fo m.neg_wtdbgp0k17.ctg.fa    # for consensus
```

Script for running second Redbean

```
#!/bin/bash -e

#SBATCH --nodes 1
#SBATCH --cpus-per-task 1
#SBATCH --ntasks 10
#SBATCH --partition=large
#SBATCH --job-name wtdbg2_2.1
#SBATCH --mem=50G
#SBATCH --time=48:00:00
#SBATCH --account=uoo02752
#SBATCH --output=%x_%j.out
#SBATCH --error=%x_%j.err
#SBATCH --mail-type=ALL
#SBATCH --mail-user=bhaup057@student.otago.ac.nz
#SBATCH --hint=nomultithread

module load wtdbg/2.5-GCC-9.2.0

wtdbg2 -x ont -g 67m -t 10 -i m.neg_ont_nanolyse_porechop.fastq.gz \
-fo m.neg_wtdbgp0k17 -p 19 -k 0 -S 2 -L 5000 --edge-min 2 —rescue-low-cov-edges -A --aln-dovetail -1     # for denovo assembly

wtpoa-cns -t 10 -i m.neg_wtdbgp19k0.ctg.lay.gz -fo m.neg_wtdbgp19k0.ctg.fa    # for consensus
```

After getting these two Redbean assembly, we will run quast to quickly check their quality and busco score and we will use the assembly with better busco score as primary and merge another in to that.

## 7. Assembly polishing with Racon
We will run 5 iteration of racon, it utilizes minimap2 to map the reads to the draft assembly to polish it.

```
#!/bin/bash -e

#SBATCH --nodes 1
#SBATCH --cpus-per-task 1
#SBATCH --ntasks 10
#SBATCH --partition=bigmem
#SBATCH --job-name racon.flye.nem
#SBATCH --mem=50G
#SBATCH --time=24:00:00
#SBATCH --account=uoo02752
#SBATCH --output=%x_%j.out
#SBATCH --error=%x_%j.err
#SBATCH --mail-type=ALL
#SBATCH --mail-user=bhaup057@student.otago.ac.nz
#SBATCH --hint=nomultithread

module load Racon
module load minimap2

#======> Correction 1 X Racon
minimap2 -t 10 assembly.fasta m.neg_ont_nanolyse_porechop.fastq > m.neg.flye.gfa1.paf
racon -t 10 m.neg_ont_nanolyse_porechop.fastq m.neg.flye.gfa1.paf assembly.fasta > m.neg.flye.racon1.fasta

#======> Correction 2 X Racon
minimap2 -t 10 m.neg.flye.racon1.fasta m.neg_ont_nanolyse_porechop.fastq > m.neg.flye.gfa2.paf
racon -t 10 m.neg_ont_nanolyse_porechop.fastq m.neg.flye.gfa2.paf m.neg.flye.racon1.fasta > m.neg.flye.racon2.fasta

#======> Correction 3 X Racon
minimap2 -t 10 m.neg.flye.racon2.fasta m.neg_ont_nanolyse_porechop.fastq > m.neg.flye.gfa3.paf
racon -t 10 m.neg_ont_nanolyse_porechop.fastq m.neg.flye.gfa3.paf m.neg.flye.racon2.fasta > m.neg.flye.racon3.fasta

#======> Correction 4 X Racon
minimap2 -t 10 m.neg.flye.racon3.fasta m.neg_ont_nanolyse_porechop.fastq > m.neg.flye.gfa4.paf
racon -t 10 m.neg_ont_nanolyse_porechop.fastq m.neg.flye.gfa4.paf m.neg.flye.racon3.fasta > m.neg.flye.racon4.fasta

#======> Correction 5 X Racon
#minimap2 -t 10 m.neg.flye.racon4.fasta m.neg_ont_nanolyse_porechop.fastq > m.neg.flye.gfa5.paf
#racon -t 10 m.neg_ont_nanolyse_porechop.fastq m.neg.flye.gfa5.paf m.neg.flye.racon4.fasta > m.neg.flye.racon5.fasta
```




